File: 섹션06. DFS - 경우의 수 만들기/04-01. 순열 (permutation), BOJ 15649 - N과 M (1).cpp
#include <stdio.h>

#define MAX (8 + 3)

int N, M;

int num_of_cases[MAX];
bool visit[MAX]; 

void printCases()
{
	for (int i = 0; i < M; i++) printf("%d ", num_of_cases[i]);
	putchar('\n');
}

void DFS(int depth)
{
	if (depth == M)
	{
		printCases();
		return;
	}

	for (int i = 1; i <= N; i++)
	{
		if (visit[i] == true) continue;

		num_of_cases[depth] = i;

		visit[i] = true;
		DFS(depth + 1);		
		visit[i] = false;
	}
}

int main()
{
	scanf("%d %d", &N, &M);

	DFS(0);

	return 0;
}

File: Python/섹션06. DFS - 경우의 수 만들기/04-01. 순열 (permutation), BOJ 15649 - N과 M (1).py
MAX = 8 + 3

N, M = map(int, input().split())

num_of_cases = [0] * MAX
visit = [False] * MAX

def print_cases():
    print(*num_of_cases[:M])

def DFS(depth):
    if depth == M:
        print_cases()
        return
    
    for i in range(1, N + 1):
        if visit[i] == True : continue
        
        num_of_cases[depth] = i
        
        visit[i] = True
        DFS(depth + 1)
        visit[i] = False
                 
DFS(0)